### PR TITLE
Add support for suspending ANCM startup

### PR DESCRIPTION
**Pull request by pakrym**
Uses the `ASPNETCORE_STARTUP_SUSPEND_EVENT` environment variable to get an event name.

Fixes: https://github.com/aspnet/AspNetCore/issues/6972

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp [hunk 1]
```diff
@@ -133,6 +133,23 @@ APPLICATION_INFO::CreateApplication(IHttpContext& pHttpContext)
...
+
+        HandleWrapper<NullHandleTraits> eventHandle = OpenEvent(SYNCHRONIZE, false, startupEvent.value().c_str());
+
+        if (eventHandle == nullptr)
+        {
+            LOG_INFOF(L"Unable to open startup suspend event");
+        }
+        else
+        {
+            LOG_LAST_ERROR_IF(WaitForSingleObject(eventHandle, INFINITE) != WAIT_OBJECT_0);
```
Comment (jkotalik) [inline src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: Thoughts on having a timeout here? What are the downsides if this hangs forever?

Comment (pakrym) [reply src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: I don't think there are any downsides, that's what ANCM was literally asked to do.

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs [hunk 2]
```diff
@@ -520,6 +521,40 @@ public async Task SetCurrentDirectoryHandlerSettingWorks()
...
+            deploymentParameters.EnvironmentVariables["ASPNETCORE_STARTUP_SUSPEND_EVENT"] = "ANCM_TestEvent";
+
+            var eventWaitHandle = new  EventWaitHandle(false, EventResetMode.ManualReset, "ANCM_TestEvent");
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var request = deploymentResult.HttpClient.GetAsync("/HelloWorld");
+
+            // didn't figure out a better way to check that ANCM is waiting to start
+            var applicationDll = Path.Combine(deploymentResult.ContentRoot, "InProcessWebSite.dll");
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs]: I don't have anything better. This is what I did for checking if a file was in use, which is similar to this. https://github.com/aspnet/AspNetCore/blob/master/src/Servers/IIS/IIS/test/testassets/OutOfProcessWebSite/Startup.cs#L71

**Comment by pakrym** [comment]
:up::date: improved the test.

File: src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp [hunk 3]
```diff
@@ -133,6 +133,30 @@ APPLICATION_INFO::CreateApplication(IHttpContext& pHttpContext)
...
+
+        if (eventHandle == nullptr)
+        {
+            LOG_INFOF(L"Unable to open startup suspend event");
+        }
+        else
+        {
+            auto const suspendedEventName = startupEvent.value() + L"_suspended";
+
+            HandleWrapper<NullHandleTraits> suspendedEventHandle = OpenEvent(EVENT_MODIFY_STATE, false, suspendedEventName.c_str());
```
Comment (jkotalik) [inline src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: Is this event only used for testing?

Comment (jkotalik) [reply src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: Feels like we shouldn't open an event if it purely used for testing.

Comment (pakrym) [reply src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: Why not? Makes for better testing, caught a test bug on IIS. And I don't think it hurts anything.

**Review by jkotalik — APPROVED** [review]

**Comment by BillHiebert** [comment]
Note that the way it currently works is that there are two events. One ANCM waits on before processing the request, and another ANCM signals to indicate it is going to process a request.

**Comment by pakrym** [comment]
Why is the second event required? Isn't request processing going to happen right after the first event is signalled?

**Pull request merged**